### ISSUE TITLE
Using Local backend, localization of relative symbolic links doesn't work

### ISSUE THREAD

**Issue by delocalizer** [broadinstitute/cromwell#1950]
Not sure if this is an actual bug or in the category of "you shouldn't expect that to work"...

given:
```
/mnt/lustre/home/conradL/foo
                         ├── bar
                         ├── baz -> bar
                         └── quux -> /mnt/lustre/home/conradL/foo/bar
```
The following WDL works fine when the input specifies `/mnt/lustre/home/conradL/foo/quux`, i.e. a symbolic link to absolute path, but not when it specifies `/mnt/lustre/home/conradL/foo/baz`, i.e. a link to a relative path:
```
workflow symLinkTest {
  call referenceTheLink { input:
    #f1="/mnt/lustre/home/conradL/foo/quux"
    f1="/mnt/lustre/home/conradL/foo/baz"
  }
}

task referenceTheLink {
  File f1
  command {}
}
```

In the second case it dies with:
```
[2017-02-03 15:49:01,37] [error] WorkflowManagerActor Workflow 6cf6c785-dc48-4409-bbd1-6f1411211f42 failed (during ExecutingWorkflowState): Failures during localization:
Could not localize /mnt/lustre/home/conradL/foo/baz -> /mnt/lustre/home/conradL/cromwell-executions/symLinkTest/6cf6c785-dc48-4409-bbd1-6f1411211f42/call-referenceTheLink/inputs/mnt/lustre/home/conradL/foo/baz:
	/mnt/lustre/home/conradL/bar doesn't exists
	File not found /mnt/lustre/home/conradL/cromwell-executions/symLinkTest/6cf6c785-dc48-4409-bbd1-6f1411211f42/call-referenceTheLink/inputs/mnt/lustre/home/conradL/foo/baz -> /mnt/lustre/home/conradL/bar
	File not found /mnt/lustre/home/conradL/bar
	File not found /mnt/lustre/home/conradL/bar
cromwell.backend.sfs.SharedFileSystem$$anonfun$localizeInputs$1$$anon$1: Failures during localization:
Could not localize /mnt/lustre/home/conradL/foo/baz -> /mnt/lustre/home/conradL/cromwell-executions/symLinkTest/6cf6c785-dc48-4409-bbd1-6f1411211f42/call-referenceTheLink/inputs/mnt/lustre/home/conradL/foo/baz:
	/mnt/lustre/home/conradL/bar doesn't exists
	File not found /mnt/lustre/home/conradL/cromwell-executions/symLinkTest/6cf6c785-dc48-4409-bbd1-6f1411211f42/call-referenceTheLink/inputs/mnt/lustre/home/conradL/foo/baz -> /mnt/lustre/home/conradL/bar
	File not found /mnt/lustre/home/conradL/bar
	File not found /mnt/lustre/home/conradL/bar
	at cromwell.backend.sfs.SharedFileSystem$$anonfun$localizeInputs$1.applyOrElse(SharedFileSystem.scala:200)
	at cromwell.backend.sfs.SharedFileSystem$$anonfun$localizeInputs$1.applyOrElse(SharedFileSystem.scala:199)
	at scala.runtime.AbstractPartialFunction.apply(AbstractPartialFunction.scala:36)
	at scala.util.Failure.recoverWith(Try.scala:203)
	at cromwell.backend.sfs.SharedFileSystem$class.localizeInputs(SharedFileSystem.scala:199)
	at cromwell.backend.sfs.SharedFileSystemJobCachingActorHelper$$anon$1.localizeInputs(SharedFileSystemJobCachingActorHelper.scala:40)
	at cromwell.backend.sfs.SharedFileSystemAsyncJobExecutionActor$$anonfun$commandLinePreProcessor$1.apply(SharedFileSystemAsyncJobExecutionActor.scala:127)
	at cromwell.backend.sfs.SharedFileSystemAsyncJobExecutionActor$$anonfun$commandLinePreProcessor$1.apply(SharedFileSystemAsyncJobExecutionActor.scala:127)
	at cromwell.backend.wdl.Command$.instantiate(Command.scala:27)
	at cromwell.backend.standard.StandardAsyncExecutionActor$class.instantiatedCommand(StandardAsyncExecutionActor.scala:83)
	at cromwell.backend.impl.sfs.config.BackgroundConfigAsyncJobExecutionActor.instantiatedCommand$lzycompute(ConfigAsyncJobExecutionActor.scala:113)
	at cromwell.backend.impl.sfs.config.BackgroundConfigAsyncJobExecutionActor.instantiatedCommand(ConfigAsyncJobExecutionActor.scala:113)
	at cromwell.backend.sfs.SharedFileSystemAsyncJobExecutionActor$class.execute(SharedFileSystemAsyncJobExecutionActor.scala:136)
	at cromwell.backend.impl.sfs.config.BackgroundConfigAsyncJobExecutionActor.execute(ConfigAsyncJobExecutionActor.scala:113)
	at cromwell.backend.standard.StandardAsyncExecutionActor$$anonfun$executeOrRecover$2.apply(StandardAsyncExecutionActor.scala:306)
	at cromwell.backend.standard.StandardAsyncExecutionActor$$anonfun$executeOrRecover$2.apply(StandardAsyncExecutionActor.scala:300)
	at scala.util.Try$.apply(Try.scala:192)
	at cromwell.backend.standard.StandardAsyncExecutionActor$class.executeOrRecover(StandardAsyncExecutionActor.scala:300)
	at cromwell.backend.impl.sfs.config.BackgroundConfigAsyncJobExecutionActor.executeOrRecover(ConfigAsyncJobExecutionActor.scala:113)
	at cromwell.backend.async.AsyncBackendJobExecutionActor$$anonfun$cromwell$backend$async$AsyncBackendJobExecutionActor$$robustExecuteOrRecover$1.apply(AsyncBackendJobExecutionActor.scala:47)
	at cromwell.backend.async.AsyncBackendJobExecutionActor$$anonfun$cromwell$backend$async$AsyncBackendJobExecutionActor$$robustExecuteOrRecover$1.apply(AsyncBackendJobExecutionActor.scala:47)
	at cromwell.core.retry.Retry$.withRetry(Retry.scala:37)
	at cromwell.backend.async.AsyncBackendJobExecutionActor$class.withRetry(AsyncBackendJobExecutionActor.scala:43)
	at cromwell.backend.async.AsyncBackendJobExecutionActor$class.cromwell$backend$async$AsyncBackendJobExecutionActor$$robustExecuteOrRecover(AsyncBackendJobExecutionActor.scala:47)
	at cromwell.backend.async.AsyncBackendJobExecutionActor$$anonfun$receive$1.applyOrElse(AsyncBackendJobExecutionActor.scala:71)
	at scala.PartialFunction$OrElse.applyOrElse(PartialFunction.scala:171)
	at akka.actor.Actor$class.aroundReceive(Actor.scala:484)
	at cromwell.backend.impl.sfs.config.BackgroundConfigAsyncJobExecutionActor.aroundReceive(ConfigAsyncJobExecutionActor.scala:113)
	at akka.actor.ActorCell.receiveMessage(ActorCell.scala:526)
	at akka.actor.ActorCell.invoke(ActorCell.scala:495)
	at akka.dispatch.Mailbox.processMailbox(Mailbox.scala:257)
	at akka.dispatch.Mailbox.run(Mailbox.scala:224)
	at akka.dispatch.Mailbox.exec(Mailbox.scala:234)
	at scala.concurrent.forkjoin.ForkJoinTask.doExec(ForkJoinTask.java:260)
	at scala.concurrent.forkjoin.ForkJoinPool$WorkQueue.runTask(ForkJoinPool.java:1339)
	at scala.concurrent.forkjoin.ForkJoinPool.runWorker(ForkJoinPool.java:1979)
	at scala.concurrent.forkjoin.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:107)
```

**Comment by Horneth** [comment]
I *think* this is the intended behaviour when resolving a symlink with a relative target. 
This answer has an example: http://unix.stackexchange.com/a/105645
TL;DR when resolving the target of the symlink, it resolves it relatively to the current directory, not the directory the source link was in.

**Comment by cjllanwarne** [comment]
I would say that when you ask Cromwell to localise a symbolic link, you would more often expect the dereferenced **file** to be localised, rather than the **link**.

**Comment by delocalizer** [comment]
@Horneth I don't think that's correct that this is intended behavior, from the OS anyway. The text quoted in that stackexchange answer in fact suggests to me exactly the behaviour I expect, but don't see:  

> "...a relative link is interpreted in relation to _its parent directory_"
not "the current directory".

If it were "the current directory" then trying to reference a sibling relative symlink would always fail unless I happened to be in the actual directory that the link target existed in, but that's not the case: 
If I have:
```
[conradL@hpcpbs01 ~]$ pwd
/mnt/lustre/home/conradL
[conradL@hpcpbs01 ~]$ tree foo
foo
├── bar
├── baz -> bar
└── quux -> /mnt/lustre/home/conradL/foo/bar
```
then `cat foo/baz` executed from my home directory is just fine:
```
[conradL@hpcpbs01 ~]$ cat foo/baz
bar
```

This is also what I expected from cromwell, - i.e. to follow the links until the file is found (like `readlink -f`)

**Comment by Horneth** [comment]
@delocalizer Hey sorry I forgot to give you an update on that and I don't know if you've seen it. You were right and the bug should be fixed now. It was due to a bug in the `better files` library we use and I filed a ticket on their github https://github.com/pathikrit/better-files/issues/115.
In the meantime I added a workaround.

**Comment by delocalizer** [comment]
Thanks @thib, for the update and explanation and workaround fix!


________________________________
